### PR TITLE
Add Support for Vault Raft Storage Backend

### PR DESCRIPTION
**Pull request by sakiib**
Signed-off-by: sakibalamin <sakibalamin@appscode.com>

### CODE REVIEW

**Comment by tks98** [comment]
Is there any update on if/when this will be merged? Being able to use this operator with Vault utilizing integrated raft storage would be hugely beneficial to my team. Thank you!

File: pkg/vault/storage/raft/raft.go [hunk 1]
```diff
@@ -17,121 +17,173 @@ limitations under the License.
...
+				},
+			)
+		}
+	}
 
-	// TODO Configure the PVCs
+	for idx := range pt.Spec.Containers {
+		if pt.Spec.Containers[idx].Name == string(api.VaultServerServiceVault) {
+			pt.Spec.Containers[idx].VolumeMounts = append(pt.Spec.Containers[idx].VolumeMounts, core.VolumeMount{
+				Name:      "data",
```
Comment (kamolhasan) [inline pkg/vault/storage/raft/raft.go:110]: const VautlVolumeMountData = "data"

File: pkg/vault/storage/raft/raft.go [hunk 2]
```diff
@@ -17,121 +17,173 @@ limitations under the License.
...
+				Secret: &core.SecretVolumeSource{
+					SecretName: o.vs.GetCertSecretName(string(api.VaultStorageCert)),
+				},
+			},
+		})
+
+		for idx := range pt.Spec.Containers {
+			if pt.Spec.Containers[idx].Name == string(api.VaultServerServiceVault) {
+				pt.Spec.Containers[idx].VolumeMounts = append(pt.Spec.Containers[idx].VolumeMounts, core.VolumeMount{
+					Name:      raftTLSAssetVolume,
```
Comment (kamolhasan) [inline pkg/vault/storage/raft/raft.go:129]: const VautlServerVolumeMountStorageCert = "storage-cert"

File: pkg/vault/storage/raft/raft.go [hunk 3]
```diff
@@ -17,121 +17,173 @@ limitations under the License.
...
-		},
-	})
+	klog.Infoln("Generating storage config for raft backend")
 
-	if _, err := t.Parse(configTemplate); err != nil {
-		return "", errors.Wrap(err, "compile storage template failed")
+	if o.Path != "" {
+		params = append(params, fmt.Sprintf(`path = "%s"`, o.Path))
+	}
+	if o.NodeID != "" {
```
Comment (kamolhasan) [inline pkg/vault/storage/raft/raft.go:149]: We don't this one. Taken from env.

File: pkg/vault/storage/raft/raft.go [hunk 4]
```diff
@@ -17,121 +17,173 @@ limitations under the License.
...
+	}
+
+	// Generate & Insert the retryJoin stanza(s)
+	replicas := pointer.Int32(o.vs.Spec.Replicas)
+	if replicas == 0 {
+		replicas = 1
+	}
+	for id := 0; id < int(replicas); id++ {
+		var retryJoin []string
+		// e.g: retryJoin = append(retryJoin, fmt.Sprint(` leader_api_addr = "https://vault-0.vault-internal.demo.svc:8200"`))
```
Comment (kamolhasan) [inline pkg/vault/storage/raft/raft.go:169]: Use more relevant comment

File: pkg/vault/storage/raft/raft.go [hunk 5]
```diff
@@ -17,121 +17,173 @@ limitations under the License.
...
+		retryJoin = append(retryJoin, fmt.Sprintf(` leader_api_addr = "%s://%s-%d.%s.%s.svc:8200"`, o.vs.Scheme(), o.vs.Name, id, o.vs.ServiceName(api.VaultServerServiceInternal), o.vs.Namespace))
+		if o.vs.Spec.TLS != nil {
+			retryJoin = append(retryJoin, fmt.Sprintf(` leader_ca_cert_file = "%s"`, filepath.Join(RaftTLSAssetDir, RaftClientCaName)))
```
Comment (kamolhasan) [inline pkg/vault/storage/raft/raft.go:172]: Use vs.CertificateMountPath() helper method

File: pkg/vault/unsealer/aws/aws_kms_ssm.go [hunk 6]
```diff
@@ -34,11 +34,13 @@ const (
 
 type Options struct {
 	api.AwsKmsSsmSpec
+	Backend string
```
Comment (kamolhasan) [inline pkg/vault/unsealer/aws/aws_kms_ssm.go:37]: remove unused params

File: pkg/vault/util/vault_config.go [hunk 7]
```diff
@@ -38,42 +40,48 @@ const (
...
 // ListenerConfig creates tcp listener config
-func GetListenerConfig() string {
-	listenerCfg := fmt.Sprintf(listenerFmt,
-		filepath.Join(VaultTLSAssetDir, core.TLSCertKey),
-		filepath.Join(VaultTLSAssetDir, core.TLSPrivateKeyKey),
-		filepath.Join(VaultTLSAssetDir, conapi.TLSCACertKey))
+func GetListenerConfig(isTLSEnabled bool) string {
+	var params []string
+	if isTLSEnabled {
+		params = append(params, fmt.Sprintf(`tls_cert_file = "%s"`, filepath.Join(VaultTLSAssetDir, core.TLSCertKey)))
```
Comment (kamolhasan) [inline pkg/vault/util/vault_config.go:58]: Use vs.CertificateMountPath() helper method

**Review by kamolhasan — CHANGES_REQUESTED** [review]

File: pkg/vault/storage/raft/raft.go [hunk 8]
```diff
@@ -17,121 +17,168 @@ limitations under the License.
...
 }
 
+// VaultRaftVolumeName is the name of the backend volume created.
+const VaultRaftVolumeName = "vault-raft-backend"
+
+// TLS related file name for Raft
 const (
-	// VaultRaftVolumeName is the name of the backend volume created.
-	VaultRaftVolumeName = "vault-raft-backend"
+	RaftCertificatePath = "/etc/vault/tls/"
```
Comment (kamolhasan) [inline pkg/vault/storage/raft/raft.go:47]: Remove this one

File: pkg/vault/storage/raft/raft.go (hunk 1 above)
Comment (kamolhasan) [reply pkg/vault/storage/raft/raft.go:108]: VautlVolumeMountConfig = "config"

**Review by kamolhasan — CHANGES_REQUESTED** [review]

**Review by kamolhasan — DISMISSED** [review]
LGTM!

**Review by kamolhasan — APPROVED** [review]

**Pull request merged**